extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js

```diff
@@ -1013,6 +1051,96 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
   }
+
+  /** @override */
+  getNonAmpCreativeRenderingMethod(headerValue) {
+    return this.isFluid_ ? XORIGIN_MODE.SAFEFRAME :
+        super.getNonAmpCreativeRenderingMethod(headerValue);
+  }
+
+  /** @override */
+  getAdditionalContextMetadata() {
+    const attributes = /** @type {!JsonObject} */ ({});
+    if (this.isFluid_) {
+      attributes['uid'] = 1;
+      attributes['hostPeerName'] = this.win.location.origin;
+      // The initial geometry isn't used for anything important, but it is
+      // expected, so we pass this string with all zero values.
+      attributes['initialGeometry'] = JSON.stringify(
+          /** @type {!JsonObject} */ ({
+            'windowCoords_t': 0,
+            'windowCoords_r': 0,
+            'windowCoords_b': 0,
+            'windowCoords_l': 0,
+            'frameCoords_t': 0,
+            'frameCoords_r': 0,
+            'frameCoords_b': 0,
+            'frameCoords_l': 0,
+            'styleZIndex': 'auto',
+            'allowedExpansion_t': 0,
+            'allowedExpansion_r': 0,
+            'allowedExpansion_b': 0,
+            'allowedExpansion_l': 0,
+            'xInView': 0,
+            'yInView': 0,
+          }));
+      attributes['permissions'] = JSON.stringify(
+          /** @type {!JsonObject} */ ({
+            expandByOverlay: false,
+            expandByPush: false,
+            readCookie: false,
+            writeCookie: false,
+          }));
+      attributes['metadata'] = JSON.stringify(
+          /** @type {!JsonObject} */ ({
+            shared: {
+              'sf_ver': this.safeframeVersion,
+              'ck_on': 1,
+              'flash_ver': '26.0.0',
+            },
+          }));
+      attributes['reportCreativeGeometry'] = true;
+      attributes['isDifferentSourceWindow'] = false;
+      attributes['sentinel'] = this.sentinel;
+    }
+    return attributes;
+  }
+
+  /** @override */
+  getXdomainCreativeFrameMessageListeners() {
+    return !this.isFluid_ ? {} : {
+      'creative_geometry_update': data => {
+        // The first creative_geometry_update message will contain bad
+        // geometric data, as it will have been computed using the initial,
+        // incorrect, iframe style. We use this first message as a
+        // triggering point to restyle the iframe, which will in turn cause
+        // safeframe to send another creative_geometry_update message, this
+        // time containing the correct information.
+        const styleString = this.iframe.getAttribute('style');
+        if (/width: 0px/.test(styleString) &&
+            /height: 0px/.test(styleString)) {
+          setStyles(this.iframe, {
+            width: '100%',
+            height: '100%',
+            position: 'relative',
+          });
+        } else {
+          const payload = tryParseJson(data['p']);
+          if (payload['width'] == 0) {
+            return;
+          }
+          this.attemptChangeSize(payload['height'], undefined)
```

Comment:
Are there any constraints around height values?  What happens if its NaN or negative?

Reply:
I trusted safeframe implicitly to not give me bad input, but I'll add a guard.

Reply:
Actually, there's not much we can do on this front since safeframe returns units with the number. If what safeframe returns is garbage, we'll just execute the catch and collapse the slot as we would with an explicit check anyway, so I think this is fine.